sycl/test/matrix/matrix-tf32-test.cpp

```diff
@@ -0,0 +1,165 @@
+  buffer<precision::tf32, 2> bufB(B.get_data(), range<2>(K, N));
+  buffer<float, 2> bufC((float *)C.get_data(), range<2>(M, N));
+
+  queue q;
+  q.submit([&](handler &cgh) {
+     auto accC = bufC.get_access<access::mode::read_write>(cgh);
+     auto accA = bufA.get_access<access::mode::read_write>(cgh);
+     auto accB = bufB.get_access<access::mode::read_write>(cgh);
+
+     cgh.parallel_for<class imatrix>(
+         nd_range<2>({NDRangeM, NDRangeN * SG_SZ}, {1, 1 * SG_SZ}),
+         [accA, accB, accC, M, N, K](nd_item<2> spmd_item)
+             [[intel::reqd_sub_group_size(SG_SZ)]]
+
+         {
+           // The submatrix API has to be accessed by all the workitems in a
+           // subgroup these functions will be called once by the subgroup no
+           // code divergence between the workitems
+           const auto global_idx = spmd_item.get_global_id(0);
+           const auto global_idy = spmd_item.get_global_id(1);
+           const auto sg_startx = global_idx - spmd_item.get_local_id(0);
+           const auto sg_starty = global_idy - spmd_item.get_local_id(1);
+
+           sub_group sg = spmd_item.get_sub_group();
+           joint_matrix<precision::tf32, TM, TK> sub_a(sg);
+           joint_matrix<precision::tf32, TK, TN, matrix_layout::packed_b> sub_b(
+               sg);
+           joint_matrix<float, TM, TN> sub_c(sg);
+           joint_matrix_load(sg, sub_c,
+                             accC.get_pointer() + (sg_startx * TM) * N +
+                                 sg_starty / SG_SZ * TN,
+                             N, matrix_layout::row_major);
+           for (int k = 0; k < K; k += TK) {
+             joint_matrix_load(sg, sub_a,
+                               accA.get_pointer() + (sg_startx * TM) * K + k, K,
+                               matrix_layout::row_major);
+             // Assume we alreay in vnni format.
+             joint_matrix_load(sg, sub_b,
+                               accB.get_pointer() + (k) * (N) +
+                                   sg_starty / SG_SZ * TN,
+                               N, matrix_layout::packed_b);
+             sub_c = joint_matrix_mad(sg, sub_a, sub_b, sub_c);
+           }
+           auto wi_slice_a = sub_a.get_wi_data();
+           for (int i = 0; i < wi_slice_a.length(); i++) {
+             float elem = wi_slice_a[i];
+             // TODO: OP= is buggy: __spirv_VectorInsertDynamic should take
+             // storage element type as argument
+             // wi_slice_a[i] *= 2;
+           }
+           joint_matrix_store(sg, sub_c,
+                              accC.get_pointer() + (sg_startx * TM) * N +
+                                  sg_starty / SG_SZ * TN,
+                              N, matrix_layout::row_major);
+         }); // parallel for
+   }).wait();
+}
+
+static constexpr size_t MATRIX_M = TM * 2;
+static constexpr size_t MATRIX_N = TN * 2;
+static constexpr size_t MATRIX_K = TK * 2;
+precision::tf32 A[MATRIX_M][MATRIX_K];
```

Comment:
In https://github.com/intel/llvm/pull/5870/files#diff-34520e7c212ec666342a649b7448a9841062a5e93eca4958990ac1653568f0d5R82
I am using the free functions defined in https://github.com/intel/llvm/pull/5870/files#diff-f71a436bdeda598b29caad471fa637a2844a12f38fe4e85b15b2ccb37bd09833R607 which accept and return floats, since floats are to be used as the fragment type for tf32. 

Perhaps this is suitable instead of making the function return the empty class.

Reply:
@hdelan, thanks, I will correct that. I made it this way so I don't change the load API for this draft PR. I will change the load API and keep the buffers as floats.
Yes the expectation is to use the free conversion function using the element indexing as argument to perform the conversion. The only issue is that we should be using get_wi_data function to get WI portion as it exhibits better semantics than just the member "data" that can confuse the user thinking is this the SG matrix or the WI portion.